vowpalwabbit/cb_explore_adf_rnd.cc

```diff
@@ -0,0 +1,307 @@
+// Copyright (c) by respective owners including Yahoo!, Microsoft, and
+// individual contributors. All rights reserved. Released under a BSD (revised)
+// license as described in the file LICENSE.
+
+#include "cb_explore_adf_rnd.h"
+#include "reductions.h"
+#include "cb_adf.h"
+#include "rand48.h"
+#include "bs.h"
+#include "gd_predict.h"
+#include "gen_cs_example.h"
+#include "cb_explore.h"
+#include "explore.h"
+#include <vector>
+#include <algorithm>
+#include <cmath>
+
+namespace VW
+{
+namespace cb_explore_adf
+{
+namespace rnd
+{
+struct cb_explore_adf_rnd
+{
+ private:
+  float epsilon;
+  float alpha;
+  float sqrtinvlambda;
+  uint32_t numrnd;
+
+  size_t increment;
+  vw* all;
+
+ public:
+  cb_explore_adf_rnd(float _epsilon, float _alpha, float _invlambda, size_t _numrnd, size_t _increment, vw* _all) : epsilon(_epsilon), alpha(_alpha), sqrtinvlambda(std::sqrt(_invlambda)), numrnd(_numrnd), increment(_increment), all(_all) { }
+  ~cb_explore_adf_rnd() = default;
+
+  // Should be called through cb_explore_adf_base for pre/post-processing
+  void predict(VW::LEARNER::multi_learner& base, multi_ex& examples) { predict_or_learn_impl<false>(base, examples); }
+  void learn(VW::LEARNER::multi_learner& base, multi_ex& examples) { predict_or_learn_impl<true>(base, examples); }
+
+ private:
+  template <bool is_learn>
+  void predict_or_learn_impl(VW::LEARNER::multi_learner& base, multi_ex& examples);
+
+  std::vector<float> bonuses;
+  std::vector<float> initials;
```

Comment:
It looks like you've structured the class as:
- private members
- public methods
- private methods

These private members should probably be in the first block right?

Reply:
Done